File: frontend/js/nomina.js
(() => {
  const state = { empleados: [], periodo: null };

  function periodoActual() {
    const hoy = new Date();
    return `${hoy.getFullYear()}-${String(hoy.getMonth() + 1).padStart(2, "0")}`;
  }

  async function init() {
    renderShell("nomina.html", "Nómina");
    try {
      state.empleados = await api.get("/empleados");
    } catch (err) {
      handleApiError(err);
    }
    fillPeriodos();
    loadResumen();
    loadNomina();
    bindFiltros();
    bindNuevaNomina();
  }

  function fillPeriodos() {
    const select = document.getElementById("f-periodo");
    const actual = periodoActual();
    const opciones = [];
    const [y, m] = actual.split("-").map(Number);
    for (let i = 0; i < 6; i++) {
      const date = new Date(y, m - 1 - i, 1);
      const val = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}`;
      opciones.push(val);
    }
    select.innerHTML = opciones.map((p) => `<option value="${p}">${formatMonthPeriodo(p)}</option>`).join("");
    state.periodo = actual;
    select.value = actual;
  }

  function etiquetaPago(quincena) {
    if (quincena === 1) return "1.ª quincena";
    if (quincena === 2) return "2.ª quincena";
    return "Mes completo";
  }

  function bindFiltros() {
    document.getElementById("f-periodo").addEventListener("change", (e) => {
      state.periodo = e.target.value;
      loadResumen();
      loadNomina();
    });
  }

  async function loadResumen() {
    const host = document.getElementById("nomina-kpis");
    try {
      const resumen = await api.get(`/nomina/resumen?periodo=${state.periodo}`);
      host.innerHTML = `
        <div class="kpi-card fade-up">
          <div class="kpi-label">Neto pagado a trabajadores</div>
          <div class="kpi-value" id="kpi-total">$0</div>
          <div class="kpi-sub">${formatMonthPeriodo(state.periodo)}</div>
        </div>
        <div class="kpi-card fade-up">
          <div class="kpi-label">Costo real para la empresa</div>
          <div class="kpi-value" id="kpi-costo">$0</div>
          <div class="kpi-sub">Carga prestacional: +${resumen.carga_prestacional}% sobre el neto</div>
        </div>
        <div class="kpi-card fade-up">
          <div class="kpi-label">Próximo pago</div>
          <div class="kpi-value" style="font-size:20px;">${resumen.proximo_pago ? formatDate(resumen.proximo_pago) : "—"}</div>
          <div class="kpi-sub">${escapeHtml(resumen.proximo_pago_concepto) || "Fecha estimada de desembolso"}</div>
          <div class="kpi-sub">${resumen.empleados_mensuales} mensual(es) · ${resumen.empleados_quincenales} quincenal(es)</div>
        </div>
        <div class="kpi-card fade-up">
          <div class="kpi-label">Novedades sin procesar</div>
          <div class="kpi-value" id="kpi-novedades">0</div>
          <div class="kpi-sub">Pueden afectar la nómina del período</div>
        </div>
      `;
      animateNumber(document.getElementById("kpi-total"), resumen.nomina_total_mes, { money: true });
      animateNumber(document.getElementById("kpi-costo"), resumen.costo_total_empleador, { money: true });
      animateNumber(document.getElementById("kpi-novedades"), resumen.novedades_sin_procesar);
    } catch (err) {
      handleApiError(err);
    }
  }

  async function loadNomina() {
    const tbody = document.getElementById("nomina-tbody");
    tbody.innerHTML = `<tr><td colspan="9" class="table-empty">Cargando nómina…</td></tr>`;
    try {
      const registros = await api.get(`/nomina?periodo=${state.periodo}`);
      renderTable(registros);
    } catch (err) {
      handleApiError(err);
      tbody.innerHTML = `<tr><td colspan="9" class="table-empty">No se pudo cargar la nómina.</td></tr>`;
    }
  }

  function renderTable(registros) {
    const tbody = document.getElementById("nomina-tbody");
    if (!registros.length) {
      tbody.innerHTML = `<tr><td colspan="9" class="table-empty">No hay registros de nómina para este período.</td></tr>`;
      return;
    }
    tbody.innerHTML = registros
      .map(
        (n) => `
      <tr>
        <td>
          <div class="person-name">${escapeHtml(n.empleado_nombre)}</div>
          ${n.novedades_mes > 0 ? `<span class="person-sub">${n.novedades_mes} novedad(es) este mes</span>` : ""}
        </td>
        <td>
          <span class="badge ${n.quincena ? "badge-info" : "badge-neutral"}">${etiquetaPago(n.quincena)}</span>
          <div class="person-sub">${n.fecha_pago ? formatDate(n.fecha_pago) : "—"} · ${n.dias_liquidados} días</div>
        </td>
        <td>
          ${formatMoney(n.salario_base)}
          ${n.quincena ? `<div class="person-sub">Devengado ${formatMoney(n.salario_devengado)}</div>` : ""}
        </td>
        <td>
          ${formatMoney(n.auxilio_transporte + n.auxilio_movilidad)}
          <div class="person-sub">Transporte ${formatMoney(n.auxilio_transporte)} · Movilidad ${formatMoney(n.auxilio_movilidad)}</div>
        </td>
        <td>
          ${formatMoney(n.total_descuentos)}
          <div class="person-sub">Salud ${formatMoney(n.salud_empleado)} · Pensión ${formatMoney(n.pension_empleado)}</div>
        </td>
        <td><strong>${formatMoney(n.total)}</strong></td>
        <td>
          ${formatMoney(n.total_prestaciones)}
          <div class="person-sub">Prima ${formatMoney(n.prima)} · Cesantías ${formatMoney(n.cesantias)} · Vac. ${formatMoney(n.provision_vacaciones)} · Pensión ${formatMoney(n.pension_empleador)} · ARL ${formatMoney(n.arl)}</div>
        </td>
        <td><strong>${formatMoney(n.costo_empleador)}</strong></td>
        <td>
          <span class="badge ${n.pagada ? "badge-success" : "badge-neutral"}">${n.pagada ? "Pagada" : "Pendiente"}</span>
          <button class="list-item-remove write-only" data-eliminar="${n.id}" style="margin-left:6px;" title="Eliminar registro">
            <svg width="15" height="15" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><polyline points="3 6 5 6 21 6"/><path d="M19 6l-1 14a2 2 0 0 1-2 2H8a2 2 0 0 1-2-2L5 6"/></svg>
          </button>
        </td>
      </tr>`
      )
      .join("");

    tbody.querySelectorAll("[data-eliminar]").forEach((btn) => {
      btn.addEventListener("click", async () => {
        if (!confirm("¿Eliminar este registro de nómina?")) return;
        try {
          await api.del(`/nomina/${btn.dataset.eliminar}`);
          showToast("Registro eliminado");
          loadNomina();
          loadResumen();
        } catch (err) {
          handleApiError(err);
        }
      });
    });
  }

  function bindNuevaNomina() {
    document.getElementById("btn-nueva-nomina").addEventListener("click", () => {
      const overlay = openModal(`
        <div class="modal-header">
          <h3>Registrar nómina</h3>
          <button class="icon-btn" data-close-modal>
            <svg width="18" height="18" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><line x1="18" y1="6" x2="6" y2="18"/><line x1="6" y1="6" x2="18" y2="18"/></svg>
          </button>
        </div>
        <form id="form-nomina">
          <div class="modal-body">
            <div class="field">
              <label>Empleado</label>
              <select name="empleado_id" required>
                <option value="">Selecciona un empleado</option>
                ${state.empleados.map((e) => `<option value="${e.id}" data-periodicidad="${e.periodicidad_pago}">${escapeHtml(e.nombre_completo)} — ${e.periodicidad_pago}</option>`).join("")}
              </select>
            </div>
            <div class="field-row">
              <div class="field">
                <label>Período (AAAA-MM)</label>
                <input type="month" name="periodo" value="${state.periodo}" required>
              </div>
              <div class="field">
                <label>Salario base</label>
                <input type="number" name="salario_base" min="0" required>
                <span class="text-faint" style="font-size:11px;">Salario mensual del contrato, aunque el pago sea quincenal</span>
              </div>
            </div>
            <div class="field">
              <label>Período de pago</label>
              <select name="quincena" id="f-quincena">
                <option value="">Mes completo (se paga el último día del mes)</option>
                <option value="1">Primera quincena (se paga el 15)</option>
                <option value="2">Segunda quincena (se paga el último día del mes)</option>
              </select>
              <span class="text-faint" style="font-size:11px;" id="hint-quincena">Se ajusta automáticamente según la periodicidad del empleado.</span>
            </div>
            <div class="field-row">
              <div class="field">
                <label>Auxilio de movilidad</label>
                <input type="number" name="auxilio_movilidad" min="0" value="0">
                <span class="text-faint" style="font-size:11px;">Lo define Ecodes para roles de campo</span>
              </div>
              <div class="field">
                <label>Otros descuentos</label>
                <input type="number" name="descuentos" min="0" value="0">
                <span class="text-faint" style="font-size:11px;">Préstamos, embargos, etc.</span>
              </div>
            </div>
            <div class="field">
              <label>Auxilio de transporte</label>
              <input type="number" name="auxilio_transporte" min="0" value="0">
              <span class="text-faint" style="font-size:11px;">Déjalo en 0 y el sistema aplica el valor de ley ($249.095) si el salario no supera 2 SMLMV ($3.501.810)</span>
            </div>
            <p class="text-faint" style="font-size:12px;">Las deducciones de salud y pensión (4% cada una) y las prestaciones que asume la empresa (prima, cesantías e intereses, vacaciones, pensión del empleador y ARL) se calculan automáticamente.</p>
          </div>
          <div class="modal-footer">
            <button type="button" class="btn btn-secondary" data-close-modal>Cancelar</button>
            <button type="submit" class="btn btn-primary">Registrar nómina</button>
          </div>
        </form>
      `);
      const selEmpleado = overlay.querySelector('select[name="empleado_id"]');
      const selQuincena = overlay.querySelector("#f-quincena");
      const hintQuincena = overlay.querySelector("#hint-quincena");
      selEmpleado.addEventListener("change", () => {
        const opcion = selEmpleado.selectedOptions[0];
        const periodicidad = opcion ? opcion.dataset.periodicidad : "";
        if (periodicidad === "Quincenal") {
          selQuincena.value = "1";
          hintQuincena.textContent = "Este empleado cobra quincenal: registra una quincena por cada pago.";
        } else if (periodicidad === "Mensual") {
          selQuincena.value = "";
          hintQuincena.textContent = "Este empleado cobra mensual: un solo registro por mes.";
        }
      });

      overlay.querySelector("#form-nomina").addEventListener("submit", async (ev) => {
        ev.preventDefault();
        const fd = new FormData(ev.target);
        const payload = {
          empleado_id: Number(fd.get("empleado_id")),
          periodo: fd.get("periodo"),
          salario_base: Number(fd.get("salario_base")),
          auxilio_transporte: Number(fd.get("auxilio_transporte") || 0),
          auxilio_movilidad: Number(fd.get("auxilio_movilidad") || 0),
          descuentos: Number(fd.get("descuentos") || 0),
          quincena: fd.get("quincena") ? Number(fd.get("quincena")) : null,
        };
        try {
          await api.post("/nomina", payload);
          showToast("Nómina registrada correctamente");
          closeModal();
          loadNomina();
          loadResumen();
        } catch (err) {
          handleApiError(err);
        }
      });
    });
  }

  init();
})();

File: backend/app/routers/nomina.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.deps import get_current_user, require_write
from app.models import Empleado, EstadoEmpleado, Nomina, Novedad, PeriodicidadPago
from app.schemas import NominaCreate, NominaOut, NominaResumen
from app.utils import fecha_de_pago, liquidar_nomina, novedades_del_mes, proximo_pago

router = APIRouter(prefix="/nomina", tags=["Nómina"])


def _to_out(nomina: Nomina, db: Session) -> NominaOut:
    data = NominaOut.model_validate(nomina)
    data.empleado_nombre = nomina.empleado.nombre_completo
    data.novedades_mes = novedades_del_mes(nomina.empleado, nomina.periodo)
    return data


@router.get("", response_model=list[NominaOut])
def listar_nomina(
    periodo: str | None = None,
    empleado_id: int | None = None,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    query = db.query(Nomina).options(joinedload(Nomina.empleado))
    if periodo:
        query = query.filter(Nomina.periodo == periodo)
    if empleado_id:
        query = query.filter(Nomina.empleado_id == empleado_id)
    registros = query.order_by(Nomina.periodo.desc()).all()
    return [_to_out(n, db) for n in registros]


@router.get("/resumen", response_model=NominaResumen)
def resumen_nomina(
    periodo: str | None = None, db: Session = Depends(get_db), current_user=Depends(get_current_user)
):
    periodo = periodo or date.today().strftime("%Y-%m")
    registros = db.query(Nomina).filter(Nomina.periodo == periodo).all()
    total = sum(float(n.total) for n in registros)
    costo_empleador = sum(float(n.costo_empleador) for n in registros)
    sin_procesar = db.query(Novedad).filter(Novedad.procesada.is_(False)).count()

    activos = db.query(Empleado).filter(Empleado.estado == EstadoEmpleado.activo).all()
    quincenales = [e for e in activos if e.periodicidad_pago == PeriodicidadPago.quincenal]
    mensuales = [e for e in activos if e.periodicidad_pago == PeriodicidadPago.mensual]

    siguiente, concepto = proximo_pago(len(mensuales), len(quincenales))

    return NominaResumen(
        nomina_total_mes=round(total, 2),
        costo_total_empleador=round(costo_empleador, 2),
        carga_prestacional=round((costo_empleador / total - 1) * 100, 1) if total else 0,
        proximo_pago=siguiente.isoformat(),
        proximo_pago_concepto=concepto,
        empleados_mensuales=len(mensuales),
        empleados_quincenales=len(quincenales),
        novedades_sin_procesar=sin_procesar,
    )


@router.post("", response_model=NominaOut, status_code=status.HTTP_201_CREATED)
def crear_nomina(
    payload: NominaCreate, db: Session = Depends(get_db), current_user=Depends(require_write)
):
    empleado = db.query(Empleado).filter(Empleado.id == payload.empleado_id).first()
    if empleado is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Empleado no encontrado")

    # Si no se indica la quincena, se usa la periodicidad configurada al empleado:
    # a quien cobra quincenalmente se le registra por defecto la primera quincena.
    quincena = payload.quincena
    if quincena is None and empleado.periodicidad_pago == PeriodicidadPago.quincenal:
        quincena = 1

    liq = liquidar_nomina(
        empleado,
        payload.salario_base,
        payload.auxilio_transporte,
        payload.auxilio_movilidad,
        payload.descuentos,
        quincena=quincena,
    )

    nomina = Nomina(
        empleado_id=payload.empleado_id,
        periodo=payload.periodo,
        quincena=quincena,
        fecha_pago=fecha_de_pago(payload.periodo, quincena),
        dias_liquidados=liq.dias_liquidados,
        salario_devengado=liq.salario_devengado,
        salario_base=liq.salario_base,
        auxilio_transporte=liq.auxilio_transporte,
        auxilio_movilidad=liq.auxilio_movilidad,
        salud_empleado=liq.salud_empleado,
        pension_empleado=liq.pension_empleado,
        descuentos=liq.otros_descuentos,
        total_descuentos=liq.total_descuentos,
        total=liq.neto_pagado,
        prima=liq.prima,
        cesantias=liq.cesantias,
        intereses_cesantias=liq.intereses_cesantias,
        provision_vacaciones=liq.provision_vacaciones,
        pension_empleador=liq.pension_empleador,
        arl=liq.arl,
        otros_aportes=liq.otros_aportes,
        total_prestaciones=liq.total_prestaciones,
        costo_empleador=liq.costo_empleador,
        pagada=payload.pagada,
    )
    db.add(nomina)
    db.commit()
    db.refresh(nomina)
    return _to_out(nomina, db)


@router.delete("/{nomina_id}", status_code=status.HTTP_204_NO_CONTENT)
def eliminar_nomina(
    nomina_id: int, db: Session = Depends(get_db), current_user=Depends(require_write)
):
    nomina = db.query(Nomina).filter(Nomina.id == nomina_id).first()
    if nomina is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Registro de nómina no encontrado")
    db.delete(nomina)
    db.commit()
